### PR TITLE
Fix memory leak for size-zero ndarray

### LINKED ISSUE #13951: Gluon RNN memory leaks with extra variables
Note: Providing complete information in the most concise form is the best way to get help. This issue template serves as the checklist for essential information to most of the technical issues and bug reports. For non-technical issues and feature requests, feel free to present the information in what you believe is the best form.

For Q & A and discussion, please start a discussion thread at https://discuss.mxnet.io 

## Description
Gluon allows one to define extra variables that may not lead to model outcome. However, having them may cause memory leak.

## Environment info (Required)

```
----------Python Info----------
Version      : 3.6.5
Compiler     : GCC 7.2.0
Build        : ('default', 'Apr 29 2018 16:14:56')
Arch         : ('64bit', '')
------------Pip Info-----------
Version      : 10.0.1
Directory    : /home/ec2-user/anaconda3/envs/mxnet_p36/lib/python3.6/site-packages/pip
----------MXNet Info-----------
Version      : 1.3.1
Directory    : /home/ec2-user/anaconda3/envs/mxnet_p36/lib/python3.6/site-packages/mxnet
Commit Hash   : 19c501680183237d52a862e6ae1dc4ddc296305b
----------System Info----------
Platform     : Linux-4.14.77-70.82.amzn1.x86_64-x86_64-with-glibc2.9
system       : Linux
node         : ip-172-16-95-144
release      : 4.14.77-70.82.amzn1.x86_64
version      : #1 SMP Mon Dec 3 20:01:27 UTC 2018
----------Hardware Info----------
machine      : x86_64
processor    : x86_64
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                8
On-line CPU(s) list:   0-7
Thread(s) per core:    2
Core(s) per socket:    4
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 79
Model name:            Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
Stepping:              1
CPU MHz:               2706.669
BogoMIPS:              4600.11
Hypervisor vendor:     Xen
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              46080K
NUMA node0 CPU(s):     0-7
----------Network Test----------
Setting timeout: 10
Timing for MXNet: https://github.com/apache/incubator-mxnet, DNS: 0.0020 sec, LOAD: 1.0198 sec.
Timing for Gluon Tutorial(en): http://gluon.mxnet.io, DNS: 0.0912 sec, LOAD: 0.1530 sec.
Timing for Gluon Tutorial(cn): https://zh.gluon.ai, DNS: 0.5845 sec, LOAD: 0.1434sec.
Timing for FashionMNIST: https://apache-mxnet.s3-accelerate.dualstack.amazonaws.com/gluon/dataset/fashion-mnist/train-labels-idx1-ubyte.gz, DNS: 0.0089 sec, LOAD: 0.1170 sec.
Timing for PYPI: https://pypi.python.org/pypi/pip, DNS: 0.0100 sec, LOAD: 0.3888 sec.
Timing for Conda: https://repo.continuum.io/pkgs/free/, DNS: 0.0104 sec, LOAD: 0.0782 sec.```
```

Package used (Python/R/Scala/Julia): Python

## Error Message:
If you run `watch -n0.1 nvidia-smi`, you may observe memory growth every by 2MB every few seconds.

## Minimum reproducible example
See [mxnet-memory-leak.tar.gz](https://github.com/apache/incubator-mxnet/files/2780496/mxnet-memory-leak.tar.gz)
The main differences between the attachment and `examples/gluon/language_model/` are to add `extra` on Line 56 in `model.py` add to add `mx.nd.array([], ctx=context)` on Line 166 and 183 in `train.py`

## Steps to reproduce
(Paste the commands you ran that produced the error.)
```
1. python train.py --cuda --tied --nhid 200 --emsize 200 --epochs 20 --dropout 0.2 &
2. watch -n0.1 nvidia-smi
```

## What have you tried to solve it?

1. Add a dummy link between all inputs and outputs. However, this may not always be possible / convenient / readable.
2. I previously suggested a feature request to allow `None` input types in the `gluon` models. Communicated with @szha that this would not be fundamentally challenging. However, this has not been acted upon and may be a low-hanging fruit alongside the memory fix leak.

Related: https://github.com/apache/incubator-mxnet/issues/13247

### LINKED ISSUE #14358: Memory builds up when creating size-zero NDArray in a loop
Note: Providing complete information in the most concise form is the best way to get help. This issue template serves as the checklist for essential information to most of the technical issues and bug reports. For non-technical issues and feature requests, feel free to present the information in what you believe is the best form.

For Q & A and discussion, please start a discussion thread at https://discuss.mxnet.io 

## Description
Memory builds up when creating size-zero ndarray in a loop

## Environment info (Required)

```
----------Python Info----------
('Version      :', '2.7.15')
('Compiler     :', 'GCC 7.2.0')
('Build        :', ('default', 'May  1 2018 23:32:55'))
('Arch         :', ('64bit', ''))
------------Pip Info-----------
('Version      :', '10.0.1')
('Directory    :', '/home/ubuntu/anaconda3/envs/python2/lib/python2.7/site-packages/pip')
----------MXNet Info-----------
('Version      :', '1.4.0')
('Directory    :', '/home/ubuntu/anaconda3/envs/python2/lib/python2.7/site-packages/mxnet')
('Commit Hash   :', 'a03d59ed867ba334d78d61246a1090cd1868f5da')
----------System Info----------
('Platform     :', 'Linux-4.4.0-1075-aws-x86_64-with-debian-stretch-sid')
('system       :', 'Linux')
('node         :', 'ip-172-31-4-52')
('release      :', '4.4.0-1075-aws')
('version      :', '#85-Ubuntu SMP Thu Jan 17 17:15:12 UTC 2019')
----------Hardware Info----------
('machine      :', 'x86_64')
('processor    :', 'x86_64')
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                32
On-line CPU(s) list:   0-31
Thread(s) per core:    2
Core(s) per socket:    16
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 79
Model name:            Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
Stepping:              1
CPU MHz:               2699.984
CPU max MHz:           3000.0000
CPU min MHz:           1200.0000
BogoMIPS:              4600.09
Hypervisor vendor:     Xen
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              46080K
NUMA node0 CPU(s):     0-31
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single kaiser fsgsbase bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx xsaveopt
----------Network Test----------
Setting timeout: 10
Timing for MXNet: https://github.com/apache/incubator-mxnet, DNS: 0.0021 sec, LOAD: 0.6245 sec.
Timing for PYPI: https://pypi.python.org/pypi/pip, DNS: 0.0071 sec, LOAD: 0.3581 sec.
Timing for FashionMNIST: https://apache-mxnet.s3-accelerate.dualstack.amazonaws.com/gluon/dataset/fashion-mnist/train-labels-idx1-ubyte.gz, DNS: 0.0265 sec, LOAD: 0.0987 sec.
Timing for Conda: https://repo.continuum.io/pkgs/free/, DNS: 0.0080 sec, LOAD: 0.0543 sec.
Timing for Gluon Tutorial(en): http://gluon.mxnet.io, DNS: 0.1285 sec, LOAD: 0.1622 sec.
Timing for Gluon Tutorial(cn): https://zh.gluon.ai, DNS: 0.3537 sec, LOAD: 0.3427 sec.

```

Package used (Python/R/Scala/Julia): Python

## Error Message:
If you run watch -n5 nvidia-smi, you may observe memory growth every by 2MB every few seconds.

## Minimum reproducible example
```
import mxnet as mx
import time

count = 0
while True:
    a = mx.nd.array([], ctx=mx.gpu(0))
    a.asnumpy()
    time.sleep(0.01)
    count += 1
    if count % 1000 == 0:
        print(count)
```

## Steps to reproduce
(Paste the commands you ran that produced the error.)

1. Put the above code into a file called leak.py
2. python leak.py
3. watch -n5 nvidia-smi

## What have you tried to solve it?

1. Create non size-zero ndarray (e.g. mx.nd.array([1], ctx=mx.gpu(1))) in the loop and there is no memory builds-up issue. But the issue remains with size-zero ndarray

Related to #13951

### PR DESCRIPTION
**Pull request by yuxihu**
Fixes #13951 Fixes #14358 

For size-zero ndarray (e.g. mx.nd.array([]), mx.nd.ones(0)), the storage handle size is 0. Currently we only free handles which size is larger than 0. This leads to memory leak for size-zero ndarray.

In this PR, we remove the check on storage handle size which was used to decide if we need to free a storage handle. After relaxing the check, we need to make sure nullptr is not reused in pooled storage manager and the context for aux handle is correctly set for sparse ndarray. 

With this PR, the memory leak issues mentioned above are fixed.

### CODE REVIEW

**Comment by yuxihu** [comment]
@yifeim @apeforest @anirudh2290 please review.

@mxnet-label-bot add [pr-awaiting-review]

**Comment by vandanavk** [comment]
@yuxihu Please add "Fixes https://github.com/apache/incubator-mxnet/issues/14358" as well, so that #14358 is also closed when this PR is merged.

**Comment by access2rohit** [comment]
@yuxihu Can you look at the failing checks. LGTM

**Comment by yuxihu** [comment]
Yes, I am looking into the test failures.

**Comment by lupesko** [comment]
Very nice catch @yuxihu !

File: include/mxnet/ndarray.h [hunk 1]
```diff
@@ -986,9 +986,8 @@ class NDArray {
 #endif
         delay_alloc = false;
       } else if (shandle.size < dbytes) {
-        // free storage if necessary and alloc again
-        if (shandle.size > 0) Storage::Get()->Free(shandle);
-        // init storage
```
Comment (access2rohit) [inline include/mxnet/ndarray.h]: nit: deleted previous comment for initializing storage after freeing it.

File: include/mxnet/ndarray.h [hunk 2]
```diff
@@ -1055,9 +1054,12 @@ class NDArray {
       }
       size_t aux_bytes = shape.Size() * mshadow::mshadow_sizeof(aux_types[i]);
       if (aux_handles[i].size < aux_bytes) {
-        // free storage if necessary and alloc again
-        if (aux_handles[i].size > 0) Storage::Get()->Free(aux_handles[i]);
-        // init aux storage
```
Comment (access2rohit) [inline include/mxnet/ndarray.h]: nit: same as above

File: src/ndarray/ndarray.cc [hunk 3]
```diff
@@ -134,9 +134,8 @@ void NDArray::Chunk::CheckAndAllocData(const mxnet::TShape &shape, int dtype) {
       << "data is expected to be allocated after aux_data";
   auto dbytes = shape.Size() * mshadow::mshadow_sizeof(dtype);
   if (shandle.size < dbytes) {
-    // free storage if necessary and alloc again
-    if (shandle.size > 0) Storage::Get()->Free(shandle);
-    // init storage
```
Comment (access2rohit) [inline src/ndarray/ndarray.cc]: nit: same as above

File: src/storage/pooled_storage_manager.h [hunk 4]
```diff
@@ -155,6 +155,10 @@ void GPUPooledStorageManager::Alloc(Storage::Handle* handle) {
 }
 
 void GPUPooledStorageManager::Free(Storage::Handle handle) {
+  // Do nothing if dptr is nullptr. Otherwise, nullptr may be reused
+  // which can cause illegal memory access error.
+  if (handle.dptr == nullptr) return;
```
Comment (access2rohit) [inline src/storage/pooled_storage_manager.h]: 👍

**Comment by access2rohit** [comment]
Left few nit picky comments

File: include/mxnet/ndarray.h (hunk 1 above)
Comment (yuxihu) [reply include/mxnet/ndarray.h]: I do not quite get what you are suggesting. I merged the previous comments into one line. Basically we free first and alloc again, which is a resize operation.

**Comment by yuxihu** [comment]
@eric-haibin-lin please help review.

@mxnet-label-bot update [pr-awaiting-review]

File: include/mxnet/ndarray.h (hunk 1 above)
Comment (access2rohit) [reply include/mxnet/ndarray.h]: Ohh ... sorry if I wasn't clear. I meant that in the previous version of code there waas this comment line:
// init aux storage
before
shandle = Storage::Get()->Alloc(dbytes, shandle.ctx);
 which got deleted in your commit. Suggesting that it would be good idea to add it back. Earlier comment already mentioned "// free storage if necessary and alloc again" and still used "// init storage". Anyways its not that big a deal. Current comment still looks reasonable.

Comment (yuxihu) [reply include/mxnet/ndarray.h]: I see. The original one was not accurate. "free storage if necessary and alloc again" was not true for the Free function. I will leave as it is for now.

File: include/mxnet/ndarray.h [hunk 5]
```diff
@@ -1055,9 +1054,12 @@ class NDArray {
...
+        // set the context for aux handle to make sure we free it to the right device.
+        if (aux_handles[i].ctx.dev_type != ctx.dev_type) {
```
Comment (apeforest) [inline include/mxnet/ndarray.h]: Is this related to memory leak? Why added here?

Comment (yuxihu) [reply include/mxnet/ndarray.h]: I do not think the memory leak comes from here. But it is better to have the same behavior regarding when we can call Free.

Comment (eric-haibin-lin) [reply include/mxnet/ndarray.h]: This usually do not happen. Have you observed such mismatch? Would it be more appropriate to add CHECK_EQ instead of modifying the ctx?

Comment (yuxihu) [reply include/mxnet/ndarray.h]: Yes, test_operator_gpu:test_sparse_nd_elemwise_add failed if context is not set here. A new aux handle is created [here](https://github.com/apache/incubator-mxnet/blob/master/include/mxnet/ndarray.h#L1054) with the default context (CPU) and size(0). Currently we have the size > 0 check so we do not call Free. After we remove the size > 0 check, it will call Free with CPU context, which caused [this failure](https://github.com/apache/incubator-mxnet/blob/master/src/storage/storage.cc#L134).

**Comment by yuxihu** [comment]
> Do we still around size 0 to page size?

Yes. I am not sure the reason behind the logic so I do not change it in this PR. We also use aligned alloc in CPU which also allocates 16/64 bytes when size is 0.

**Comment by yuxihu** [comment]
@eric-haibin-lin It is ready for final review.
